src/PhpUnit/UriInterfaceTestsTrait.php

```diff
@@ -0,0 +1,290 @@
+     *   valid as PHP does not allow to throw an exception in __toString():
+     *     - If the path is rootless and an authority is present, the path MUST
+     *       be prefixed by "/".
+     *     - If the path is starting with more than one "/" and no authority is
+     *       present, the starting slashes MUST be reduced to one.
+     * - If a query is present, it MUST be prefixed by "?".
+     * - If a fragment is present, it MUST be prefixed by "#".
+     */
+    public function testToString(
+        $url,
+        $scheme,
+        $userinfo,
+        $host,
+        $port,
+        $authority,
+        $path,
+        $query,
+        $fragment,
+        $normalized
+    ) {
+        $url = $this->createUriObject($url);
+        Assert::assertSame($normalized, $url->__toString());
+    }
+
+    /**
+     * @dataProvider uriProvider
+     * @group scheme
+     */
+    public function testWithScheme($url)
+    {
+        $url = $this->createUriObject($url);
+        Assert::assertEquals($url->getScheme(), $url->withScheme($url->getScheme())->getScheme());
+    }
+
+    /**
+     * @dataProvider uriProvider
+     * @group userinfo
+     */
+    public function testWithUserInfo($url)
+    {
+        $url  = $this->createUriObject($url);
+        $res  = explode(':', $url->getUserInfo());
+        $user = array_shift($res);
+        $pass = array_shift($res);
+
+        Assert::assertEquals($url->getUserInfo(), $url->withUserInfo($user, $pass)->getUserInfo());
+    }
+
+    /**
+     * @dataProvider uriProvider
+     * @group host
+     */
+    public function testWithHost($url)
+    {
+        $url = $this->createUriObject($url);
+        Assert::assertEquals($url->getHost(), $url->withHost($url->getHost())->getHost());
+    }
+
+    /**
+     * @dataProvider uriProvider
+     * @group port
+     */
+    public function testWithPort($url)
+    {
+        $url = $this->createUriObject($url);
+        Assert::assertContains($url->withPort($url->getPort())->getPort(), [null, $url->getPort()]);
+    }
+
+    /**
+     * @dataProvider uriProvider
+     * @group path
+     */
+    public function testWithPath($url)
+    {
+        $url = $this->createUriObject($url);
+        Assert::assertSame($url->getPath(), $url->withPath($url->getPath())->getPath());
+    }
+
+    /**
+     * @dataProvider uriProvider
+     * @group query
+     */
+    public function testWithQuery($url)
+    {
+        $url = $this->createUriObject($url);
+        Assert::assertEquals($url->getQuery(), $url->withQuery($url->getQuery())->getQuery());
+    }
+
+    /**
+     * @dataProvider uriProvider
+     * @group fragment
+     */
+    public function testWithFragment($url)
+    {
+        $url = $this->createUriObject($url);
+        Assert::assertEquals($url->getFragment(), $url->withFragment($url->getFragment())->getFragment());
+    }
+
+    public function uriProvider()
+    {
+        return [
+            'URL with full components' => [
```

Comment:
I like to add a comment on top of the array describing the array format. Something like

`// Description => [input, schema, ....]`